Resizable: Fix content shrink on resize

Daniel-Garmig opened the pull request:
Fixes: #2277. 
This end up being a bug on "resizable", and as I have shown on the issue it's quite a weird behavior and it's not consistent across browsers.

The same bug applies to the resizable element and "alsoResize" elements, so I fixed it on both places.

Dialog widget uses "alsoResize" property to update dialog contents size (as seen on the initial issue). This PR fixes that as well.

I have added some tests for both "box-sizing: content-box" and "box-sizing: border-box".

melloware reviewed (approved):
awesome!

melloware commented:
@Daniel-Garmig do you know a way I can "MonkeyPatch" your fix on top of existing jQuery UI?  We do stuff like this in our app to patch other jquery UI functions but the magic is never quite the same for all patches.

https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/resources/META-INF/resources/primefaces/jquery/jquery.ui.pfextensions.js

I would like to be able to apply this patch until its accepted by jQueryUI team or (gulp) rejected.

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -244,4 +244,57 @@ QUnit.test( "nested resizable", function( assert ) {
+	element = $( "#resizable1" ).css( { overflow: "auto" } ).resizable(),
+	handle = ".ui-resizable-se";
+
+	$( "*" ).css( "box-sizing", "border-box" );
```
That's unnecessarily slow & not clean - every element is selected and has an inline style added to set `border-box`. If you want a test with global `border-box`, add it via a `<style>` tag.

The same applies to other places.

Daniel-Garmig replied:
Ok, I'll change it!

Would it be better if I append a `<style>` tag using jQuery and remove it after the tests?
Something like this:
```javascript
var style = $( "<style> * { box-sizing: border-box; } </style>" ).appendTo( "head" );
//do test
style.remove();
``` 

One more thing. I noticed that other tests (not my new ones) do the same (see https://github.com/jquery/jquery-ui/blob/main/tests/unit/resizable/options.js#L558 ). Should I change those as well?

Thanks!

mgol replied:
You can also change that test. If there are too many of them, though, you don't have to do this here.

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -244,4 +244,57 @@ QUnit.test( "nested resizable", function( assert ) {
+				borderStyle: "solid"
+			} )
+			.appendTo( "#resizable1" ),
+	element = $( "#resizable1" ).css( { overflow: "auto" } ).resizable(),
```
```suggestion
	element = $( "#resizable1" ).css( "overflow", "auto" ).resizable(),
```
Similarly in other tests.

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -244,4 +244,57 @@ QUnit.test( "nested resizable", function( assert ) {
+			.css( {
+				width: 50,
+				height: 200,
+				padding: 10,
```
Can you also set some margin to make sure that doesn't interfere?

Also, provide explicit values with the `px` suffix, not numbers. Passing numbers & depending on them getting `px` auto-appended is deprecated and disabled for most CSS properties in jQuery 4.0.0.

The properties here work (see the regex at https://github.com/jquery/jquery/blob/4.0.0-beta.2/src/css/isAutoPx.js) but this is mostly for backwards compat reasons, let's not depend on that here.

mgol commented on ui/widgets/resizable.js:
```diff
@@ -690,6 +684,21 @@ $.widget( "ui.resizable", $.ui.mouse, {
 	},
 
+	_calculateAdjustedElementDimensions: function( element ) {
+		if ( !( /content-box/ ).test( element.css( "box-sizing" ) ) ) {
```
No need for a regex test here; just compare the value directly.

mgol commented on ui/widgets/resizable.js:
```diff
@@ -690,6 +684,21 @@ $.widget( "ui.resizable", $.ui.mouse, {
+
+		var outerDimensions = this._getPaddingPlusBorderDimensions( element );
+		return {
+			height: element[ 0 ].getBoundingClientRect().height - outerDimensions.height,
+			width: element[ 0 ].getBoundingClientRect().width - outerDimensions.width
```
`getBoundingClientRect` takes transforms into consideration while the CSS `width` & `height` - as well as jQuery APIs - do not. Using `getBoundingClientRect` is not valid. We should probably add a test case where the whole content is treated via a scale transform to make sure we didn't break it.

Daniel-Garmig commented:
@mgol I have changed the method to avoid using `getBoundingClientRect()`. 
Instead I get the value using `parseFloat( element.css( "width" ) );`, but without the scrollbars. So I disable scrollbars, get the correct computed size and set scrollbars back. (This seems kind of expensive and slow to me, but idk).

I tried to optimize it, so this calculation is only done when necessary. After a resize, width and height properties will be added as inline css to the element, so I will use those when available to avoid the expensive calculation.

I added some test for resizable elements with `transform` properties set. It doesn't matter after this changes but why not.

Let me know what do you think.

mgol commented on ui/widgets/resizable.js:
```diff
@@ -690,6 +699,45 @@ $.widget( "ui.resizable", $.ui.mouse, {
+	_getElementSizeWithoutOverflow: function( element, sizeProperty ) {
+		var overflowProperty = sizeProperty === "width" ? "overflow-y" : "overflow-x";
+
+		var origOverflow = element.css( overflowProperty );
+		element.css( overflowProperty, "hidden" );
+		var elSize = parseFloat( element.css( sizeProperty ) );
```
This will trigger an extra layout per function call which is undesirable. Since it gets called twice, it's two extra layouts, which is even worse.

There are other ways to get a scrollbar size, e.g. the hack that jQuery itself employs; see https://github.com/jquery/jquery/blob/dbc9dac7aecb106b66050342ff8daf1ecdd4239f/src/css.js#L99-L110

Daniel-Garmig replied:
Thanks for the suggestion!

I have changed the calculation to use something similar to the "jQuery hack" you linked.
In this case, I used it to calculate the "theoretical" size of the element instead of scrollbar size delta.

Daniel-Garmig commented on tests/unit/resizable/core.js:
```diff
@@ -244,4 +244,142 @@ QUnit.test( "nested resizable", function( assert ) {
+QUnit.test( "Resizable with scrollbars and box-sizing: border-box", function( assert ) {
+	assert.expect( 4 );
+
+	var style = $( "<style> * { box-sizing: border-box; } </style>" ).appendTo( "head" );
```
Ok! I changed it.

There are some unit test cases on Resizable/options that appends elements to `body` and don't clear them after. See https://github.com/jquery/jquery-ui/blob/54f96eea31b21d9ecb00912261df3e5aaebf8cce/tests/unit/resizable/options.js#L15-L20
I can fix those so it appends elements to `#qunit-fixture`. Maybe on other PR?

mgol replied:
Yes, feel free to submit another PR fixing this. No need to fix it here since those tests are unrelated.

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -244,4 +244,134 @@ QUnit.test( "nested resizable", function( assert ) {
+
+	$( "<style> * { box-sizing: border-box; } </style>" ).appendTo( "#qunit-fixture" );
+
+	//Both scrollbars
```
Please start comments with a space. Please fix other comments as well:
```suggestion
	// Both scrollbars
```

mgol commented on ui/widgets/resizable.js:
```diff
@@ -381,20 +385,25 @@ $.widget( "ui.resizable", $.ui.mouse, {
 		this.offset = this.helper.offset();
 		this.position = { left: curleft, top: curtop };
 
+		var calculatedSize = undefined;
```
Please move the declaration to the single var we have above, adding it as a first entry in its own line, without the `undefined` initializer - that's our code style:
```js
		var calculatedSize,
			scroll = ( a && a === "left" ) ? "scrollLeft" : "scrollTop",
			has = false;
```

mgol commented on ui/widgets/resizable.js:
```diff
@@ -690,6 +699,50 @@ $.widget( "ui.resizable", $.ui.mouse, {
+		}
+
+		// Check if CSS inline styles are set and use those (usually from previous resizes)
+		var elWidth = parseFloat( ce.style.width );
+		var elHeight = parseFloat( ce.style.height );
+
+		var paddingBorder = this._getPaddingPlusBorderDimensions( element );
```
In this repo, we either use a single `var` (as I described in my comment above) or multiple ones in consecutive lines, not spread within the function. This is because the JS engine will hoist the declaration anyway and we want to make it explicit. You're free to assign the values here, but the declarations should be at the top.

Variables declared without initializers should be the first on the list, preferably all in one line if they fit.

mgol commented on ui/widgets/resizable.js:
```diff
@@ -690,6 +699,50 @@ $.widget( "ui.resizable", $.ui.mouse, {
+			0.5
+
+		// If offsetWidth/offsetHeight is unknown, then we can't determine theoretical size.
+		// Use an explicit zero to avoid NaN (gh-3964)
```
`gh-3964` here is a reference to a jQuery issue, this is invalid in this repository. Please use the full link here:
```suggestion
		// Use an explicit zero to avoid NaN.
		// See https://github.com/jquery/jquery/issues/3964
```

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -244,4 +244,134 @@ QUnit.test( "nested resizable", function( assert ) {
 	outer.remove();
 } );
 
+QUnit.test( "Resizable with scrollbars and box-sizing: border-box", function( assert ) {
```
We have 4 tests here that are largely copy-pasted. I'd consider moving the test body to a generic function accepting the `borderBox` & `transforms` boolean parameters and then use them inside to apply the correct logic & messages.

See an example of a similar pattern in the jQuery repo:
https://github.com/jquery/jquery/blob/4b7ecbad24463c875f03ef4c7a7d307a091f93fd/test/unit/ajax.js#L1821-L1871

jQuery here uses an anonymous function inside the iterators to define the test body. I think I'd prefer to extract it into a parametrized function here to avoid the large indentation, but other than that it looks fine.

mgol reviewed (changes requested):
A few more comments.

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -246,41 +246,29 @@ QUnit.test( "nested resizable", function( assert ) {
 
 QUnit.test( "Resizable with scrollbars and box-sizing: content-box", function( assert ) {
 	assert.expect( 4 );
+	testResizableWithBoxSizing( assert, false, false );
```
See https://ariya.io/2011/08/hall-of-api-shame-boolean-trap; let's use named parameters for readability:
```suggestion
	testResizableWithBoxSizing( assert, {
		isBorderBox: false,
		applyScaleTransform: false
	} );
```
These are tests, we don't need to optimize for size or extreme performance.

mgol commented on tests/unit/resizable/core.js:
```diff
@@ -246,41 +246,29 @@ QUnit.test( "nested resizable", function( assert ) {
+function testResizableWithBoxSizing( assert, isBorderBox, applyScaleTransform ) {
+	var widthBefore, heightBefore,
+		cssBoxSizing = isBorderBox ? "border-box" : "content-box",
+		cssTrasform = applyScaleTransform ? "scale(1.5)" : "",
```
A typo:
```suggestion
		cssTransform = applyScaleTransform ? "scale(1.5)" : "",
```

mgol reviewed (changes requested):
Almost ready! Just two more comments from me.

Daniel-Garmig commented:
Changes done!
@mgol Thank you for all your suggestions and comments, they were great!

mgol reviewed (approved):
Thanks, looks good!

fnagel reviewed (approved):
+1 by reading and by running the unit tests (can confirm the new test will fail in Chrome if the fix is not present)

mgol commented:
Landed, thanks!

mgol commented:
jQuery UI 1.14.1 with this change has been released.
